src/nvim/eval/decode.c

```diff
@@ -0,0 +1,945 @@
+              case 'f': {
+                len++;
+                p++;
+                break;
+              }
+              default: {
+                EMSG2(_("E474: Unknown escape sequence: %s"), p - 1);
+                goto json_decode_string_fail;
+              }
+            }
+          } else {
+            uint8_t p_byte = (uint8_t) *p;
+            // unescaped = %x20-21 / %x23-5B / %x5D-10FFFF
+            if (p_byte < 0x20) {
+              EMSG2(_("E474: ASCII control characters cannot be present "
+                      "inside string: %s"), p);
+              goto json_decode_string_fail;
+            }
+            const int ch = utf_ptr2char((char_u *) p);
+            // All characters above U+007F are encoded using two or more bytes
+            // and thus cannot possibly be equal to *p. But utf_ptr2char({0xFF,
+            // 0}) will return 0xFF, even though 0xFF cannot start any UTF-8
+            // code point at all.
+            if (ch >= 0x80 && p_byte == ch) {
+              EMSG2(_("E474: Only UTF-8 strings allowed: %s"), p);
+              goto json_decode_string_fail;
+            } else if (ch > 0x10FFFF) {
+              EMSG2(_("E474: Only UTF-8 code points up to U+10FFFF "
+                      "are allowed to appear unescaped: %s"), p);
+              goto json_decode_string_fail;
+            }
+            const size_t ch_len = (size_t) utf_char2len(ch);
+            assert(ch_len == (size_t) (ch ? utf_ptr2len((char_u *) p) : 1));
+            len += ch_len;
+            p += ch_len;
+          }
+        }
+        if (*p != '"') {
+          EMSG2(_("E474: Expected string end: %s"), buf);
+          goto json_decode_string_fail;
+        }
+        if (len == 0) {
+          POP(((typval_T) {
+            .v_type = VAR_STRING,
+            .vval = { .v_string = NULL },
+          }), false);
+          break;
+        }
+        char *str = xmalloc(len + 1);
+        int fst_in_pair = 0;
+        char *str_end = str;
+        bool hasnul = false;
+#define PUT_FST_IN_PAIR(fst_in_pair, str_end) \
+        do { \
+          if (fst_in_pair != 0) { \
+            str_end += utf_char2bytes(fst_in_pair, (char_u *) str_end); \
+            fst_in_pair = 0; \
+          } \
+        } while (0)
+        for (const char *t = s; t < p; t++) {
+          if (t[0] != '\\' || t[1] != 'u') {
+            PUT_FST_IN_PAIR(fst_in_pair, str_end);
+          }
+          if (*t == '\\') {
+            t++;
+            switch (*t) {
+              case 'u': {
+                const char ubuf[] = { t[1], t[2], t[3], t[4] };
+                t += 4;
+                unsigned long ch;
+                vim_str2nr((char_u *) ubuf, NULL, NULL,
+                           STR2NR_HEX | STR2NR_FORCE, NULL, &ch, 4);
+                if (ch == 0) {
+                  hasnul = true;
+                }
+                if (SURROGATE_HI_START <= ch && ch <= SURROGATE_HI_END) {
+                  fst_in_pair = (int) ch;
```

Comment:
Don't we lose the first character of "\uD800\uD800" (both characters in the high surrogate range)?